text_extensions_for_pandas/array/token_span.py

```diff
@@ -267,49 +296,96 @@ class TokenSpanArray(SpanArray, TokenSpanOpMixin):
     * `self._end_tokens`: Numpy array of end offsets (1 + last token in span).
     """
 
-    @staticmethod
-    def from_char_offsets(tokens: SpanArray) -> "TokenSpanArray":
-        """
-        Convenience factory method for wrapping the character-level spans of a
-        series of tokens into single-token token-based spans.
-
-        :param tokens: character-based offsets of the tokens
-
-        :return: A TokenSpanArray containing single-token spans for each of the
-        tokens in `tokens`.
+    @classmethod
+    def create(cls, tokens: Union[SpanArray, Sequence[SpanArray]],
+               begin_tokens: Union[pd.Series, np.ndarray, Sequence[int]] = None,
+               end_tokens: Union[pd.Series, np.ndarray, Sequence[int]] = None):
         """
-        begin_tokens = np.arange(len(tokens))
-        return TokenSpanArray(tokens, begin_tokens, begin_tokens + 1)
+        Factory method for creating instances of this class.
 
-    def __init__(
-        self,
-        tokens: SpanArray,
-        begin_tokens: Union[pd.Series, np.ndarray, Sequence[int]] = None,
-        end_tokens: Union[pd.Series, np.ndarray, Sequence[int]] = None,
-    ):
-        """
         :param tokens: Character-level span information about the underlying
-        tokens.
+        tokens. Can be a single set of tokens, covering all spans, or a separate
+        `SpanArray` pointer for every span.
 
         :param begin_tokens: Array of begin offsets measured in tokens
         :param end_tokens: Array of end offsets measured in tokens
+
+        :return: A new `TokenSpanArray` object
         """
-        if not isinstance(tokens, SpanArray):
-            raise TypeError(f"Expected SpanArray as tokens but got {type(tokens)}")
         if not isinstance(begin_tokens, (pd.Series, np.ndarray, list)):
             raise TypeError(f"begin_tokens is of unsupported type {type(begin_tokens)}. "
                             f"Supported types are Series, ndarray and List[int].")
         if not isinstance(end_tokens, (pd.Series, np.ndarray, list)):
             raise TypeError(f"end_tokens is of unsupported type {type(end_tokens)}. "
                             f"Supported types are Series, ndarray and List[int].")
 
-        super().__init__(tokens.target_text, tokens.begin, tokens.end)
+        if isinstance(tokens, SpanArray):
+            if not tokens.is_single_document:
+                raise ValueError(f"Token spans come from more than one document.")
+            token_table = TokenTable.create_single(tokens)  # type: ThingTable
```

Comment:
Why do we need to create a table for tokens? Since the token positions are indices in the SpanArray, aren't the multiple docs already handled?

Reply:
Each span in the `TokenSpanArray` could be from a different tokenization of a different document. For example, you might mix spans over the SpaCy tokenization of document 1, spans over the BERT tokenization of document 1, and spans over the SpaCy documentation of document 2 in the same array. 

It's not really practical in this situation to maintain a single `SpanArray` to cover the entire universe of tokens. You don't know what documents exist in the current Python process. The implementation of `TokenSpanArray` here uses a separate `SpanArray` to represent a single tokenization of a single document. Each `TokenSpan` is tagged with a specific underlying tokenization.

The main purpose of the table is to track the set of distinct tokenizations. Keeping track of that set lets some operations run in constant time:
* detecting that every span in an array (or a slice of an array) is over the same tokens
* identifying all the distinct tokenizations present in an array of spans (for serialization)
* detecting that two copies of the same tokenization (i.e. different arrays, same spans) are the same tokenization

An alternative approach would be to store an array of pointers to `SpanArrays` (i.e. what is currently computed on-demand under the property `TokenSpan.tokens`) instead of using a table of distinct tokenizations. That approach would simplify the code at the expense of making some operations asymptotically slower. I might make that change; haven't decided yet.

Reply:
Went ahead and got rid of the token table. Now `TokenSpanArray` uses an array of pointers to `SpanArray` objects representing the tokenizations.

A few wrinkles with this new implementation: Some methods generate a `TokenTable` on demand, and I had to jump through hoops in a few places to keep NumPy from converting an array of `SpanArray`s into a 2D `ndarray`.